### DISCUSSION TITLE
Using a case class for both JSON parsing and SQL activities

### DISCUSSION THREAD

**Discussion by samikrc** [com-lihaoyi/scalasql#28] [Q&A] [answered]
I am trying to work out a case class which can be used for JSON parsing (using json/upickle) and then eventually writing to a database (using scalasql). Essentially the flow is something like this:
API (using cask) => JSON payload => parse into case class (using ujson) => save into database (using scalasql)

I have defined the case class like so, before introducing the scalasql related stuff:
```
case class Model
(
    key: String,
    name: String,
    desc: Option[String]
)

object Model
{
     // Another constructor
     def apply(...){}

     // Implicit method to read and write a Model object from/to JSON
     given modelReadWrite: ReadWriter[Model] = upickle.default.readwriter[ujson.Value].bimap[Model](
          model => write(model),
          json => ...
     )
}
```

The above construct works fine. Now I am trying to convert this case class so that I can use it for DB interaction as well. The new case class looks like this:
```
case class Model[T[_]]
(
    key: T[String],
    name: T[String],
    desc: T[Option[String]]
)
```
However when I put this:
```
object Model extends Table[Model] with Sc[Model]
{
...
}
```
Things error out. 

How do I achieve this, if it is possible? I could not find an example for this, but I would think this as a pretty common use case! Thanks for any pointer - love the libraries in the ecosystem!

Best.

**Comment by lihaoyi** [comment, marked as answer]
Can you try using a normal ScalaSql `case class Foo[T[_]]`, and define the upickle seriaizer via `implicit def rw: upickle.default.ReadWriter[Foo[Sc]] = macroRw[Foo[Sc]]`

**Comment by samikrc** [comment]
Thanks - yes that works!
A full example is also available in your documentation for Cask: https://com-lihaoyi.github.io/cask/index.html#todomvc-database-integration

**Comment by samikrc** [comment]
If you can link this example from scalasql page as well, that would be helpful for folks who are just starting off with scalasql.
